packages/flutter/lib/src/material/bottom_app_bar.dart

```diff
@@ -0,0 +1,135 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/rendering.dart';
+import 'package:flutter/widgets.dart';
+
+import 'material.dart';
+import 'scaffold.dart';
+
+// Examples can assume:
+// Widget bottomAppBarContents;
+
+/// A container that shows in the [Scaffold.bottomNavigationBar] slot, and can
+/// have a notch in which a [FloatingActionButton] is showing.
+///
+/// Typically used with a [Scaffold] and a [FloatingActionButton].
+///
+/// ## Sample code
+///
+/// ```dart
+/// new Scaffold(
+///   bottomNavigationBar: new BottomAppBar(
+///     color: Colors.white,
+///     child: bottomAppBarContents,
+///   ),
+///   floatingActionButton: new FloatingActionButton(onPressed: null),
+/// )
+/// ```
+///
+/// See also:
+///
+///   * [NotchMaker] a closure used for creating a notch in a shape.
+///   * [ScaffoldGeometry.floatingActionBarNotchMaker] the [NotchMaker] used to
+///   make a notch for the [FloatingActionButton]
+///   * [FloatingActionButton] which the [BottomAppBar] makes a notch for.
+class BottomAppBar extends StatefulWidget {
+  const BottomAppBar({
+    Key key,
+    this.child,
+    // TODO(amirh): use a default color from the theme.
+    @required this.color,
+    this.elevation: 8.0,
+  }) : super(key: key);
+
+  /// The widget below this widget in the tree.
+  ///
+  /// {@macro flutter.widgets.child}
+  final Widget child;
+
+  /// The color to paint the material.
+  final Color color;
+
+  /// The z-coordinate at which to place this bototm app bar. This controls the
+  /// size of the shadow below the bottom app bar.
+  ///
+  /// Defaults to 8, the appropriate elevation for bottom app bars.
+  final double elevation;
+
+  @override
+  State createState() => new BottomAppBarState();
+}
+
+class BottomAppBarState extends State<BottomAppBar> {
+  ValueListenable<ScaffoldGeometry> geometryListenable;
+
+  @override
+  Widget build(BuildContext context) {
+      return new PhysicalShape(
+        child: new Material(
+          type: MaterialType.transparency,
+          child: widget.child,
+        ),
+        clipper: new _BottomAppBarClipper(geometry: geometryListenable),
+        elevation: widget.elevation,
+        color: widget.color,
+    );
+  }
+
+  @override
+  void didChangeDependencies() {
```

Comment:
nit: this normally goes above the builder

Reply:
done